lib/charms/traefik_k8s/v2/ingress.py

```diff
@@ -534,12 +543,13 @@ def __init__(
         relation_name: str = DEFAULT_RELATION_NAME,
         *,
         host: Optional[str] = None,
+        ip: Optional[str] = None,
```

Comment:
Do we need both ip and and host?
Can we instead have something like
```suggestion
        host: Optional[Union[str, ipaddress.IPv4Address]] = None,  # TODO in v3, rename to address
```
- If None, use fqdn;
- If able to figure out that the "fqdn" is not valid (no cluster DNS or machine charm) then get bind address.

Reply:
The problem of merging the `ip` and `host` into one field and let the requirer to choice to provide `ip` or `host` is that the requirer charm doesn't have the information whether it's connected with the `traefik-k8s` charm of `nginx-ingress-integrator` charm. And for `nginx-ingress-inegrator` the `ip` is a must even the cluster DNS is enabled due to the limitation of Nginx ingress in Kubernetes.

Reply:
Yes, while the `ip` field is now technically nullable, it should only be set to `None` in the event of an issue with juju. For users, the `ip` field remains mandatory.

Reply:
mmmh ok, so should we do something in case it's None, like log a warning? To prevent people from passing None and think it's a regular supported use case. I'm not sure how to do it with pydantic, but we can also consider not typing it at all as `Optional[]` and only in a validator, accept a None passed as value and log that warning.

Reply:
For users, the methods to consider are `IngressPerAppRequirer.__init__` and `IngressPerAppRequirer.provide_ingress_requirements`. When a user passes `None` as the `ip` value, it defaults to the IP address from Juju. Therefore, users cannot intentionally set the `ip` to `None` using the public API. As for the data class, the `ip` field should be typed as nullable, as the provider need to act depend on whether the `ip` field is `None` or not, and the provider need to get that from the data class.
Logging is definitely a good idea, added, thanks!

Reply:
I follow Weii in his conclusion. We didn't want to have IP as mandatory field because depending on the provider this information is unnecessary, so we didn't want a charm related to Traefik to Error because it can't get its IP field. Setting it to nullable delegates the decision of what is needed to the provider, if the provider needs the IP field and this field is None, it can act accordingly (error, blocked, message, ...)
As Weii mentioned the Library user can't set this field using the public interface, so the risks there are limited

Is it ok for you @PietroPasotti ?

Reply:
@sed-i you too?